threedi_result_selection.py

```diff
@@ -115,3 +142,76 @@ def get_state_description(self):
                     'model_schematisation': file,
                     'result_directories': list
                 })
+
+    @property
+    def logged_in(self):
+        return self.username is not None and self.password is not None
+
+    def handle_download(self):
+        result_type_codes_download = [
+            'subgrid_map', 'flow-aggregate', 'id-mapping', 'logfiles']
+        selection_model = self.dialog.downloadResultTableView.selectionModel()
+        indexes = selection_model.selectedIndexes()
+        if len(indexes) != 1:
+            pop_up_info("Please select one result.")
+            return
+        row = indexes[0].row()
+        item = self.download_result_model.rows[row]
+        to_download = [
+            r for r in item.results.value if
+            r['result_type']['code'] in result_type_codes_download]
+        to_download_urls = [dl['attachment_url'] for dl in to_download]
+        log.debug(item.name.value)
+
+        # ask user where to store download
+        directory = QFileDialog.getExistingDirectory(
+            None,
+            'Choose a directory',
+            os.path.expanduser('~'),
+        )
+        if not directory:
+            return
+        self.download_directory = os.path.join(directory, item.name.value)
+
+        # For now, only work with empty directories that we create ourselves.
+        # Because the files are downloaded and processed in chunks, we cannot
+        # guarantee data integrity with existing files.
+        if os.path.exists(self.download_directory):
+            pop_up_info(
+                "The directory %s already exists." % self.download_directory)
+            return
+        log.info("Creating download directory.")
+        os.mkdir(self.download_directory)
+
+        log.debug(self.download_directory)
+
+        CHUNK_SIZE = 16 * 1024
+        # Important note: QNetworkAccessManager is asynchronous, which means
+        # the downloads are processed asynchronous using callbacks.
+        for url in to_download_urls:
+            request = QNetworkRequest(QUrl(url))
+            request.setRawHeader('username', self.username)
+            request.setRawHeader('password', self.password)
+            reply = self.network_manager.get(request)
+            # Get replies in chunks, and process them
+            reply.setReadBufferSize(CHUNK_SIZE)
+            reply.readyRead.connect(
+                self.on_single_download_ready_to_read_chunk)
+            reply.finished.connect(self.on_single_download_finished)
+        pop_up_info("Download started.")
+
+    def on_single_download_ready_to_read_chunk(self):
+        """Process a chunk of the downloaded data."""
+        reply = self.sender()
+        raw_chunk = reply.readAll()  # QByteArray
+        filename = reply.url().toString().split('/')[-1]
+        with open(os.path.join(self.download_directory, filename), 'ab') as f:
```

Comment:
Note to self: `self.download_directory` will change when a new scenario download is started. `QNetworkRequest.setAttribute` is perhaps an idea.